homeassistant/components/qwikswitch.py

```diff
@@ -136,61 +134,69 @@ def setup(hass, config):
     url = config[DOMAIN][CONF_URL]
     dimmer_adjust = config[DOMAIN][CONF_DIMMER_ADJUST]
 
-    qsusb = QSUsb(url, _LOGGER, dimmer_adjust)
+    def callback_value_changed(qsdevices, key, new): \
+            # pylint: disable=unused-argument
+        """Update entiry values based on device change."""
+        entity = hass.data[DOMAIN].get(key)
+        if entity is not None:
+            entity.schedule_update_ha_state()  # Part of Entity/ToggleEntity
+
+    session = async_get_clientsession(hass)
+    qsusb = QSUsb(url=url, dim_adj=dimmer_adjust, session=session,
+                  callback_value_changed=callback_value_changed)
 
-    def _stop(event):
+    @callback
+    def async_stop(event):  # pylint: disable=unused-argument
         """Stop the listener queue and clean up."""
         nonlocal qsusb
         qsusb.stop()
         qsusb = None
-        global QSUSB
-        QSUSB = {}
+        hass.data[DOMAIN] = {}
         _LOGGER.info("Waiting for long poll to QSUSB to time out")
 
-    hass.bus.listen(EVENT_HOMEASSISTANT_STOP, _stop)
+    hass.bus.async_listen(EVENT_HOMEASSISTANT_STOP, async_stop)
 
     # Discover all devices in QSUSB
-    devices = qsusb.devices()
-    QSUSB['switch'] = []
-    QSUSB['light'] = []
-    for item in devices:
-        if item[PQS_TYPE] == QSType.relay and (item[QS_NAME].lower()
-                                               .endswith(' switch')):
-            item[QS_NAME] = item[QS_NAME][:-7]  # Remove ' switch' postfix
-            QSUSB['switch'].append(QSSwitch(item, qsusb))
-        elif item[PQS_TYPE] in [QSType.relay, QSType.dimmer]:
-            QSUSB['light'].append(QSLight(item, qsusb))
+    yield from qsusb.update_from_devices()
+    hass.data[DOMAIN]['switch'] = []
+    hass.data[DOMAIN]['light'] = []
+    for _id, item in qsusb.devices:
+        if (item[QS_TYPE] == QSType.relay and
+                item[QSDATA][QS_NAME].lower().endswith(' switch')):
+            item[QSDATA][QS_NAME] = item[QSDATA][QS_NAME][:-7]  # Remove switch
+            new_dev = QSSwitch(_id, qsusb)
```

Comment:
Instead of instantiating the entities in the component, load the platforms via discovery and pass the needed discovery info for the entities to be instantiated in the platforms.